qcodes/instrument_drivers/Spectrum/M4i.py

```diff
@@ -0,0 +1,677 @@
+# **************************************************************************
+#
+# Driver file for M4i.44x-x8
+#
+# **************************************************************************
+#
+# QuTech
+#
+# Written by: Luka Bavdaz, Marco Tagliaferri, Pieter Eendebak
+# Also see: http://spectrum-instrumentation.com/en/m4i-platform-overview
+#
+
+#%%
+import logging
+import numpy as np
+import ctypes as ct
+from functools import partial
+from qcodes.utils.validators import Enum, Numbers, Anything
+from qcodes.instrument.base import Instrument
+from qcodes.instrument.parameter import ManualParameter
+try:
+    import pyspcm
+except ImportError:
+    raise ImportError('to use the M4i driver install the pyspcm module')
+
+#%% Helper functions
+
+# This function is taken from an example provided by Spectrum GmbH
+
+
+def szTypeToName(lCardType):
+    sName = ''
+    lVersion = (lCardType & pyspcm.TYP_VERSIONMASK)
+    if (lCardType & pyspcm.TYP_SERIESMASK) == pyspcm.TYP_M2ISERIES:
+        sName = 'M2i.%04x' % lVersion
+    elif (lCardType & pyspcm.TYP_SERIESMASK) == pyspcm.TYP_M2IEXPSERIES:
+        sName = 'M2i.%04x-Exp' % lVersion
+    elif (lCardType & pyspcm.TYP_SERIESMASK) == pyspcm.TYP_M3ISERIES:
+        sName = 'M3i.%04x' % lVersion
+    elif (lCardType & pyspcm.TYP_SERIESMASK) == pyspcm.TYP_M3IEXPSERIES:
+        sName = 'M3i.%04x-Exp' % lVersion
+    elif (lCardType & pyspcm.TYP_SERIESMASK) == pyspcm.TYP_M4IEXPSERIES:
+        sName = 'M4i.%04x-x8' % lVersion
+    else:
+        sName = 'unknown type'
+    return sName
+
+#%% Main driver class
+
+# TODO: whenever an error occurs (including validation errors) the python console needs to be restarted
+
+
+class M4i(Instrument):
+
+    '''
+    example usage for acquisition with channel 2 using an external trigger that triggers multiple times with trigger mode HIGH:
+
+    m4 = M4i(name='M4i', server_name=None)

```

Comment:
🎠  this is still relevant:

Class docstring is split, and should all be in under the class and not on the init function.
Cardid needs to be described.

See http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for an example.
Example should be in its directive.
And it would awesome to have type descriptions.